### PR TITLE
Avoid shadowing of inline assembly instructions with variables

### LINKED ISSUE #1457: Inline assembly variables names don't clash with instructions
```
assembly {
    let gas := 1
    call(gas, ...)
}
```

In this example `call(gas,...` will refer to the `gas` instruction as opposed to the variable. Similarly, `let call := 1` is accepted as valid too.

### PR DESCRIPTION
**Pull request by axic**
Fixes #1457.

### CODE REVIEW

**Comment by chriseth** [comment]
@axic I made some changes, please review.

File: test/libsolidity/InlineAssembly.cpp [hunk 1]
```diff
@@ -182,6 +182,18 @@ BOOST_AUTO_TEST_CASE(error_tag)
...
+BOOST_AUTO_TEST_CASE(inline_assembly_shadowed_instruction_declaration)
+{
+	// Error message: "Cannot use instruction names for identifier names."
+	BOOST_CHECK(!successAssemble("{ let gas := 1 }"));
+}
+
+BOOST_AUTO_TEST_CASE(inline_assembly_shadowed_instruction_assignment)
+{
+	// Error message: "Cannot use instruction names for identifier names."
+	BOOST_CHECK(!successAssemble("{ 2 =: gas }"));
```
Comment (axic) [inline test/libsolidity/InlineAssembly.cpp:194]: Why not `gas := 2`? (I know both are correct)

**Comment by axic** [comment]
I cannot approve it because I'm the original opener of the PR.

File: test/libsolidity/InlineAssembly.cpp (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/InlineAssembly.cpp:194]: oh right, let me check.

Comment (chriseth) [reply test/libsolidity/InlineAssembly.cpp:194]: Weird, the test should have failed...

Comment (axic) [reply test/libsolidity/InlineAssembly.cpp:194]: `2 =: gas` is still an assignment of 2 to `gas`, just mixing the functional and the non-functional notation as:

`2` is pushed to stack
`=: gas` pops item off the stack into variable `gas`

Comment (chriseth) [reply test/libsolidity/InlineAssembly.cpp:194]: I know, but the check is only in place for `gas := 2`: https://github.com/ethereum/solidity/pull/1606/commits/f62e269115822f791a0ad2e3699815f0725a0ef7#diff-9b7bbca0dea1e2bd9ff41750a720ffe2R104

Comment (chriseth) [reply test/libsolidity/InlineAssembly.cpp:194]: Should be fixed now.

File: libsolidity/inlineasm/AsmParser.cpp [hunk 2]
```diff
@@ -71,6 +71,8 @@ assembly::Statement Parser::parseStatement()
 		expectToken(Token::Colon);
 		assignment.variableName.location = location();
 		assignment.variableName.name = m_scanner->currentLiteral();
+		if (instructions().count(assignment.variableName.name))
+			fatalParserError("Identifier expected.");
```
Comment (axic) [inline libsolidity/inlineasm/AsmParser.cpp]: How about going for "Identifier expected, but got instruction name." and using it in the other assignment case (but not in declaration).

File: test/libsolidity/InlineAssembly.cpp [hunk 3]
```diff
@@ -182,6 +182,24 @@ BOOST_AUTO_TEST_CASE(error_tag)
...
+}
+
+BOOST_AUTO_TEST_CASE(inline_assembly_shadowed_instruction_functional_assignment)
+{
+	// Error message: "Cannot use instruction names for identifier names."
```
Comment (axic) [inline test/libsolidity/InlineAssembly.cpp:199]: Need to update comment to match actual error message.

File: libsolidity/inlineasm/AsmParser.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/inlineasm/AsmParser.cpp]: In the other the error will point at the `:=` symbol, so I would rather leave it as it is now.

**Comment by axic** [comment]
Good to go then.

**Review by chriseth — APPROVED** [review]

**Pull request merged**